qt-build/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+//! [cxx_build](https://docs.rs/cxx-build/latest/cxx_build/), or
+//! [cpp_build](https://docs.rs/cpp_build/latest/cpp_build/).
+
+use std::{
+    env,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+pub use versions::SemVer;
+
+/// Helper for build.rs scripts using Qt
+/// ```
+/// let qt_modules = vec!["Core", "Gui"]
+///     .iter()
+///     .map(|m| String::from(*m))
+///     .collect();
+/// let qtbuild = qt_build::QtBuild::new(qt_modules).expect("Could not find Qt installation");
+/// ```
+pub struct QtBuild {
+    pub version: SemVer,
+    qmake_executable: String,
+    qt_modules: Vec<String>,
+}
+
+impl QtBuild {
+    /// Search for where Qt is installed using qmake. Specify the Qt modules you are
+    /// linking with the `qt_modules` parameter, ommitting the `Qt` prefix (`"Core"`
+    /// rather than `"QtCore"`). After construction, use the [QtBuild::qmake_query]
+    /// method to get information about the Qt installation.
+    ///
+    /// The directories specified by the `PATH` environment variable are where qmake is
+    /// searched for. Alternatively, the `QMAKE` environment variable may be set to specify
+    /// an explicit path to qmake.
+    ///
+    /// If multiple major versions (for example, `5` and `6`) of Qt could be installed, set
+    /// the `QT_VERSION_MAJOR` environment variable to force which one to use. When using Cargo
+    /// as the build system for the whole build, prefer using `QT_VERSION_MAJOR` over the `QMAKE`
+    /// environment variable because it will account for different names for the qmake executable
+    /// that some Linux distributions use.
+    ///
+    /// However, when building a Rust staticlib that gets linked to C++ code by a C++ build
+    /// system, it is best to use the `QMAKE` environment variable to ensure that the Rust
+    /// staticlib is linked to the same installation of Qt that the C++ build system has
+    /// detected. With CMake, you can get this from the `Qt${QT_VERSION_MAJOR}::qmake`
+    /// target's `IMPORTED_LOCATION` property, for example:
+    /// ```
+    /// find_package(QT NAMES Qt6 Qt5 COMPONENTS Core REQUIRED)
+    /// find_package(Qt${QT_VERSION_MAJOR} COMPONENTS Core REQUIRED)
+    /// get_target_property(QMAKE Qt${QT_VERSION_MAJOR}::qmake IMPORTED_LOCATION)
+    ///
+    /// execute_process(
+    ///     COMMAND cmake -E env
+    ///         "CARGO_TARGET_DIR=${CMAKE_CURRENT_BINARY_DIR}/cargo"
+    ///         "QMAKE=${QMAKE}"
+    ///         cargo build
+    ///     WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    /// )
+    /// ```
+    #[allow(clippy::result_unit_err)]
+    pub fn new(mut qt_modules: Vec<String>) -> Result<Self, ()> {
+        if qt_modules.is_empty() {
+            qt_modules.push("Core".to_string());
+        }
+        println!("cargo:rerun-if-env-changed=QMAKE");
+        println!("cargo:rerun-if-env-changed=QT_VERSION_MAJOR");
+        fn verify_candidate(candidate: &str) -> Result<(&str, versions::SemVer), String> {
```

Comment:
having a nested function seems slightly odd to me, can it be a private function implement on `QtBuild` ? Or a closure, but even a closure means this `new` function is large, whereas moving the function out keeps each of them small.

Reply:
This is intentionally within the `new` function because it isn't needed by anything else: https://www.reddit.com/r/rust/comments/1ob8l7/declaring_functions_within_functions/

Reply:
OK, it was more a styling thing anyway of trying to keep functions short, lets go with it for now.